File: 백준/Silver/2468. 안전 영역/안전 영역.py
from collections import deque

n = int(input())
board = [list(map(int,input().split())) for _ in range(n)]

dx = [0,0,1,-1]
dy = [1,-1,0,0]

def check_land(): # 물 수위를 높여가면서 board새로 갱신
    cnt=0
    for i in range(n):
        for j in range(n): 
            if board[i][j]==0:
                cnt+=1
                continue
            board[i][j]-=1
    if cnt==n*n: # 모두 잠겼을 때
        return False
    else:
        return True

def bfs():
    vis = [[0]*n for _ in range(n)]
    count=0
    for i in range(n):
        for j in range(n):
            if not vis[i][j] and board[i][j]:
                count+=1
                vis[i][j]=1
                q = deque([(i,j)])
                while q:
                    x,y=q.popleft()
                    for dir in range(4):
                        nx = x + dx[dir]
                        ny = y + dy[dir]
                        if nx<0 or ny<0 or nx>=n or ny>=n:
                            continue
                        if vis[nx][ny] or not board[nx][ny]:
                            continue
                        vis[nx][ny]=1
                        q.append((nx,ny))
    return count

max_count=bfs()
for i in range(1,101):
    if check_land():
        max_count= max(max_count,bfs())
    else:
        break
print(max_count)

File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

public class Main {

  private static int[] dx = {0,0,1,-1};
  private static int[] dy = {1,-1,0,0};
  private static int N;
  private static int[][] area;

  public static void main(String[] args) throws Exception {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    N = Integer.parseInt(br.readLine());
    area = new int[N][N];
    for(int i = 0; i < N; i++){
      StringTokenizer st = new StringTokenizer(br.readLine());
      for(int j = 0; j < N; j++){
        area[i][j] = Integer.parseInt(st.nextToken());
      }
    }

    int max=0;
    for(int h=0; h<=100; h++){
      max = Math.max(max, bfs(h));
    }
    System.out.println(max);
  }

  private static int bfs(int h){
    boolean[][] vis = new boolean[N][N];

    int count =0;
    for(int i=0; i<N; i++){
      for(int j=0; j<N; j++){
        if(vis[i][j] || area[i][j]<=h) continue;
        count++;
        vis[i][j] = true;
        Deque<Point> queue = new ArrayDeque<>();
        queue.offer(new Point(i,j));

        while(!queue.isEmpty()){
          Point point = queue.poll();
          for(int dir=0; dir<4; dir++){
            int nx = point.x + dx[dir];
            int ny = point.y + dy[dir];
            if(isOutOfRange(nx, ny)) continue;
            if(vis[nx][ny] || area[nx][ny]<=h) continue;
            vis[nx][ny] =true;
            queue.offer(new Point(nx,ny));
          }
        }
      }
    }
    return count;
  }

  private static class Point{
    int x;
    int y;

    public Point(int x, int y) {
      this.x = x;
      this.y = y;
    }
  }

  private static boolean isOutOfRange(int x, int y){
    return x<0 || x>=N || y<0 || y>=N;
  }
}
